share/doc/wake/basic_wake_tutorial.md

```diff
@@ -0,0 +1,628 @@
+```
+def dubmid fullstring =
+  def Pair firststr laststr = 
+    explode fullstring
+    | splitAt (strlen fullstring / 2)
+  def dupli = laststr | take 1
+  cat (firststr ++ dupli ++ laststr)
+```
+
+* Takes a string and then removes the last character in a string
+```
+wake -x '"been"| explode | take 3| cat'
+```
+
+* Replaces the first charcter with the character "c"
+```
+wake -x '"bad"| explode | drop 1| prepend "c" | cat' 
+```
+### Tuple
+A tuple is collection of items of different types. `get`,`set` and `edit` are the three inbuilt methods that can be used with the tuples.
+* Below example shows the declaration of a tuple. For more information about the tuple look into wake database
+```
+def abcd = 1;"xx";1e5;Unit
+```
+### List
+A List is a sequence of items of same type. Empty list returns `Nil`. `head` returns the first element in the list. `tail` returns the rest of the list without the `head`. 
+* Example shows how to prepend the list of integers with the given value
+```
+def d = 1,2,3,Nil
+def listprepend data1 = prepend data1 d
+```
+
+* Example creates a list of given elements
+```
+def fone xx = xx,xx,xx,Nil
+```
+* Try out these examples
+```
+wake -x '"south" | explode | take 1| cat'
+wake -x '"north"| explode | tail | take 1 | cat'
+wake -x '"east" | explode| reverse | take 1 |cat'
+wake -x '"west" | explode | reverse | tail |take 1| cat'
+```
+
+* Below are the set of examples which shows how to extract specific characters from a string
+```
+def first s = s | explode | take 1| cat
+def second s = s | explode | tail | take 1 | cat
+def third s = s | explode| tail | tail | take 1 |cat
+def fourth s =  s | explode | tail | tail | tail | take 1| cat
+def last s = s| explode | reverse |take 1 | cat
+```
+
+* Try out these different set of examples and capture the result
+```
+def roll s = "{fourth s}{first s}{second s}{third s}"
+def exch s = "{second s}{first s}{third s}{fourth s}"
+```
+
+* Below example shows how to declare a list of strings and apply map function to it. Map is a predefined function that maps to the every element in the list
+```
+def words = "ache", "vile", "amid", "evil", "ogre" , Nil
+def mapping words = words | map roll
+def mapping2 words = map exch words
+```
+
+* Try out these examples and record the results
+```
+def what fullstring = fullstring| roll | roll | roll| exch |  roll
+def fb fullstring = fullstring | roll |roll | roll
+def fc fullstring = fullstring | roll | exch |fb
+def fd fullstring = fullstring | exch | fc | exch
+```
+
+* This examples convert the string "seat" -> "eats"
+```
+wake -x 'cat((tail (explode "seat")) ++(take 1(explode "seat")))'
+```
+
+### Type Inference
+* Here the paramater c has to be string since function strlen as been applied to it, and b has to be integer as it is being added and a also should be an integer because it is list of integers 
+```
+def madeup a b c =  (b + strlen c), a 
```

Comment:
this is the first example with multiple parameters. Is it true that in wake you can't specify the type of your parameters? You had a long text description that is the case, shouldn't you say that much earlier?

what does the `(b+ strlen c), a` do? Is it thus assuming that `a` is a list and this is making two lists, or what?

Reply:
Actually, `a` is a `List Integer` because `x, y` prepends element `x` to list `y`.

Reply:
i mean, I am not really asking in the comments... can you make it more clear in the text? I am still in the dark whether you can specify the type of your parameters in wake or not, and if you can't how that information is propogated

Reply:
We cannot specify them explicitly. By looking at the the operations performed on the parameters we can understand what should be the parameters